[Docs]: Add some more documentation on how to work with github.

PhilipMetzger opened the pull request:
The document roughly describes some workflows, which help new users to start with github.

@superpippo12 Discord self ping.

[Rendered](https://github.com/PhilipMetzger/jj/blob/push-075ec4ff589b45a89ccf25098371e4ff/docs/github.md)
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the branch rather than adding commits on top. Use force-push when
pushing the updated branch (`jj git push` does that automatically when you
rewrite a branch). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
-->

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have added tests to cover my changes

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+
+## Setup a ssh key
+As of December 2022 it's necessary to setup a ssh key to work with github
+projects. See [Github's Tutorial][gh]. This restriction may be lifted in the 
```
I think you forgot to add the link

PhilipMetzger replied:
Doing it at the moment.

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+# Using Jujutsu with Github Projects
+
+## Setup a ssh key
+As of December 2022 it's necessary to setup a ssh key to work with github
```
nit: s/setup/set up/ (since this is a verb)

martinvonz replied:
nit: s/github/GitHub/

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+# Using Jujutsu with Github Projects
+
+## Setup a ssh key
+As of December 2022 it's necessary to setup a ssh key to work with github
```
I thought HTTP authentication with password would also be possible since https://github.com/martinvonz/jj/pull/715. Is that not the case?

I think it's still a good idea to recommend using `ssh-agent`. What do you think about suggesting that here?

PhilipMetzger replied:
I didn't test it. But when I was pushing a update to this pr, it popped up.

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+## Setup a ssh key
+As of December 2022 it's necessary to setup a ssh key to work with github
+projects. See [Github's Tutorial][gh]. This restriction may be lifted in the 
+future, see issue #469 for more information and progress on authenticated http. 
```
GitHub markdown doesn't seem to auto-link the "#469", so I guess we'll have to make it

PhilipMetzger replied:
Done

PhilipMetzger commented on docs/github.md:
````diff
@@ -0,0 +1,41 @@
+
+
+## Basic workflows
+Jujutsu supports both a `git` and a `hg` like workflows. This allows you
+to continue using your favorite way of using the vcs. It's also possible to 
+just commit multiple changes and set the expected branch retroactively. 
+
+### Git workflow
+
+```shell script
+$ jj branch c $your-feature
+$ # Do your work
+$ jj commit -m "feat(xy): some thing"
+$ jj git push --branch $your-feature
+```
+
````
This needs work.

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+# Using Jujutsu with Github Projects
```
nit: s/Github/GitHub/

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+# Using Jujutsu with Github Projects
+
+## Setup a ssh key
```
nit: s/Setup a ssh/Set up an SSH/

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+
+## Basic workflows
+Jujutsu supports both a `git` and a `hg` like workflows. This allows you
+to continue using your favorite way of using the vcs. It's also possible to 
```
nit: s/vcs/VCS/ (or maybe s/vcs/tool/)

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+
+
+## Basic workflows
+Jujutsu supports both a `git` and a `hg` like workflows. This allows you
```
I think it's better to describe the workflows you have in mind and then maybe compare specifics to Git and Mercurial. Something like this:

> You can either do X and the to do Y (similar to how Git is typically used),
> or you can do Z (similar to how Mercurial is typically used)

That way it's clearer to the reader what you mean by Git-like and Mercurial-like.

martinvonz commented on docs/github.md:
````diff
@@ -0,0 +1,41 @@
+```shell script
+$ jj branch c $your-feature
+$ # Do your work
+$ jj commit -m "feat(xy): some thing"
````
We should add a note somewhere explaining that the branch doesn't move forward if the user were to create more commits on top.

PhilipMetzger replied:
Added.

martinvonz commented on docs/github.md:
````diff
@@ -0,0 +1,41 @@
+### Git workflow
+
+```shell script
+$ jj branch c $your-feature
````
nit: Spell this out as `jj branch create`? The "c" probably seems quite cryptic to new users

PhilipMetzger replied:
Done.

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+$ jj branch c $your-feature
+$ # Do your work
+$ jj commit -m "feat(xy): some thing"
+$ jj git push --branch $your-feature
```
Should we leave out the `--branch $your-feature` here since we push that branch anyway (since it's on `@-`)?

PhilipMetzger replied:
Done.

martinvonz commented on docs/github.md:
````diff
@@ -0,0 +1,41 @@
+$ jj git push --branch $your-feature
+```
+
+## Working in a git co-located repository
````
nit: s/git/Git/

martinvonz commented on docs/github.md:
````diff
@@ -0,0 +1,41 @@
+```
+
+## Working in a git co-located repository
+After doing `jj init --git-repository=.`, git will be in a detached HEAD state,
````
s/git-repository/git-repo/

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,41 @@
+need to export your changes to the underlying `git` repository after each commit.
+
+
+The same workflow is necessary to inform jujutsu about changes in the git repository.
```
s/jujutsu/Jujutsu/

PhilipMetzger commented on docs/github.md:
```diff
@@ -0,0 +1,76 @@
+In a main Jujutsu repository, the workflow is simplified as there's no need to 
+continiously import changes. 
+
+## Useful Revsets
+
+
+
```
Sadly I don't have any.

PhilipMetzger commented:
There are only a few things missing from this doc, but generally it's ready.

martinvonz commented:
> There are only a few things missing from this doc, but generally it's ready.

Okay, I'll clicked the "Ready for review" button :)

PhilipMetzger commented:
i'll need to tidy up the revsets section and then it's ready. If I could structure some sentences better, I'll gladly change them.

Ralith commented on docs/github.md:
```diff
@@ -0,0 +1,107 @@
+
+
+## Basic workflows
+You can either create a branch and then commit revisions on top of it (similiar 
+to how Git is typically used) or just start making changes in the repository
+(which is how Mercurial is used). 
+This allows you to continue using your favorite way of using the VCS. 
```
I find this a bit confusing, because the workflow is the same whether or not you create a branch first. It's also odd to repeatedly mention creating a branch before working when a git user's intention is generally going to be to place the new commits, not their starting point, on the branch.

joyously replied:
I think to explain it better, you might have to mention co-location of branches and whether you want to work on two separate things at once (switch between branches) or not (since you can't see them both at once with Git).

martinvonz replied:
I agree. I had a long unfinished message here with a suggestion, but that seem to have gotten lost (probably by a reboot) :(

I would suggest describing the workflow without providing different ones for Git and Mercurial because I wouldn't recommend the Git-like one anyway since it only works for a single commit. Something like this:


```shell
# Start a new commit off of `main`
$ jj new main
# Refactor some files, then add a description and start a new commit
$ jj commit -m 'refactor(foo): restructure foo()'
# Add a feature, then add a description and start a new commit
$ jj commit -m 'feat(bar): add support for bar'
# Create a branch so we can push it to GitHub
$ jj branch create bar -r @-
# Push the branch to GitHub (pushes only `bar`)
$ jj git push
```

martinvonz commented on docs/github.md:
```diff
@@ -0,0 +1,107 @@
+retroactively. 
+
+**Notable Differences to other VCS**  
+Branches do not move forward when creating commits on top. This means you should
+set the current branch to work on before starting your work. 
```
This is confusing to me. I would say that you should create the branch after finishing your work instead, since you'll have to do that anyway, so setting a branch before starting is just unnecessary. I suspect you're thinking of the case of single-commit PR, but I don't want users to get used to that case and then be confused when they create more than one commit.

joyously replied:
Both are confusing to me... 
What is "move forward" referring to a branch? What is "on top"?

Creating a branch after you work seems pointless. In my mind, branches keep your work-in-progress separate from your stable code, so you should branch first, test, debug, and commit when it's working. Only after all that should the branch be merged back to stable.

joyously replied:
@martinvonz Is the command sequence above what you do on this repo? I find this repo so hard to follow because there are pushes that eliminate previous ones, and they all have the same commit messages. I hope that's not what you are promoting.

martinvonz replied:
> Both are confusing to me... What is "move forward" referring to a branch? What is "on top"?

The both mean "up" in the direction that `jj/git/hg log` and `gitk` show the graph. In Git, if you check out a branch (e.g. `git checkout -b my-branch`) and create a commit, then `my-branch` will point to the new commit.

> 
> Creating a branch after you work seems pointless. In my mind, branches keep your work-in-progress separate from your stable code, so you should branch first, test, debug, and commit when it's working. Only after all that should the branch be merged back to stable.

I think this confusion is all about "named branches" vs. "anonymous branches". See https://www.mercurial-scm.org/wiki/BranchingExplained. I think both @PhilipMetzger and I were talking about branches as the associated name, not the topological concept. We really need a document describing anonymous branches... (we can probably copy a lot from that Mercurial page).

joyously replied:
OK, it doesn't help future readers to explain the words to me here...it has to be in the doc itself.

martinvonz replied:
> @martinvonz Is the command sequence above what you do on this repo?

Yes, that's pretty typical for me to do in this repo (and other repos I occasionally contribute to). I mean, creating commits and pushing them to a remote is pretty much how everyone interacts with any repo :)

> I find this repo so hard to follow because there are pushes that eliminate previous ones, and they all have the same commit messages.

We use what's often called "stacked diffs". Googling that led me to https://jg.gg/2018/09/29/stacked-diffs-versus-pull-requests/. So consider reading that if you're unfamiliar with that way of working. FWIW, the Git and Mercurial projects also do that. I think most developers at Google and Meta also do.

Unfortunately, GitHub has pretty poor support for stacked diffs :( There are tools like graphite.dev and reviewable.io, which I think both attempt to have better support for stacked diffs.

> I hope that's not what you are promoting.

I definitely do promote stacked diffs :)

Oh, in case you were thinking that this document (or my suggested snippet?) somehow promoted stacked diffs, I don't think it does. I had planned to write about both workflows here (that was part of the draft I lost on reboot).

joyously replied:
> We use what's often called "stacked diffs".
> I definitely do promote stacked diffs

I read the link and a lot of links from there also. I understand the concept, however, it is still very hard to follow this repo because clicking links in the notifications often gets a 404 and the commit messages are the same in each push. So if you are simply using `jj` on this repo, and describing that in this doc, you really need to state that and link to the explanations in the doc, as well as help mitigate the confusion of new commits with same message.

martinvonz replied:
Good point, @joyously. I had a section (in that comment I lost to a reboot) describing the two different workflows for addressing PR comments, depending on whether the reader's project uses stacked diffs (i.e. clean commits and force-pushing) or the usual commit-on-top workflow. I've tried to recreate , now also clarifying that this project prefers clean commits. @PhilipMetzger, can you incorporate the below (or something similar to it)?


## Addressing review comments

There are two different workflows for addressing review comments, depending on your project's preference. Many projects prefer that you address review comments by adding new commits to your branch. Some projects (such as this one) instead prefer that you keep your commits clean by rewriting them and then force-pushing [^stacked-diffs].

[^stacked-diffs]: If you're wondering why we prefer clean commits in this project, see e.g. [this blog post](https://jg.gg/2018/09/29/stacked-diffs-versus-pull-requests/).

### Adding new commits

If your project prefers that you address review comments by adding commits on top, you can do that by doing something like this:

```shell
# Create a new commit on top of the `bar` branch from above. (Feel free to rename the branch, @PhilipMetzger)
$ jj new bar
# <Address the comment by updating the code>
# Review the changes
$ jj diff
# Give the fix description and create a new working-copy commit on top
$ jj commit -m 'address a comment'
# Update the branch to point to the new commit
$ jj branch set bar -r @-
# Push it to the remote
$ jj git push
```

### Rewriting commits

If your project prefers that you keep commits clean, you can do that by doing something like this:

```
# Create a new commit on top of the second-to-last commit in
#`bar` (let's say that's where we were asked to fix something)
$ jj new bar-
# <Address the comment by updating the code>
# Review the changes
$ jj diff
# Squash the changes into the parent commit
$ jj squash
# Push the updated branch to the remote. This will automatically become a force push
$ jj git push --branch bar
```

joyously replied:
I see that the two workflows look very similar. Can the one with rewrite change the message also? (I know it's replacing the original, so the message _should_ be mostly the same, but it did change...) Also, can you mention that in general, force pushing to a remote will cause problems for everyone else (404s in GitHub, contributors having to rework something their change might be based on)?

PhilipMetzger replied:
I can suggest that. But it's actually not a workflow failure, but an issue how GitHub handles it in their backend. Because both [Phabricator](https://phabricator.com) and [Gerrit](https://www.gerritcodereview.com) are able to differentiate between multiple pushes with the same commit message.

joyously replied:
Regardless, this doc is about GitHub, so I think you should mention it.

PhilipMetzger replied:
I added this between the squash and push step. 
```bash
$ # Edit the commit message to make your changes visible to your reviewers.
$ jj describe
```

martinvonz replied:
> I added this between the squash and push step.
> 
> ```shell
> $ # Edit the commit message to make your changes visible to your reviewers.
> $ jj describe
> ```

I'd prefer to remove this part again. It's quite atypical that review comments lead to an updated commit message. Of course it happens sometimes, but it's much less than than 50% of the time in my experience (even less than 10%?).

martinvonz replied:
> Also, can you mention that in general, force pushing to a remote will cause problems for everyone else (404s in GitHub, contributors having to rework something their change might be based on)?

Note that even GitHub has its "Compare" button that shows changes compared to before a force-push. That doesn't always work (if you're unlucky, the old commit could have been GC'd), and it will also include changes caused by rebasing (I hope they'll some day show output like `jj interdiff` instead).

I also think it's out of scope for this page since we already tell the user to do whatever their project prefers. I think the reader typically doesn't have control over which workflow their project uses (they do if it's their own personal project, but then they would typically not use PRs). Still, I agree that it's fair to add a footnote about the disadvantages, just like we have a footnote about the advantages (though I added that footnote only because you pointed out that readers might be confused about our decision). Maybe it's better to add a pointer to an external page about the pros and cons instead (i.e. a less opinionated page than the one I linked to).

PhilipMetzger replied:
> I also think it's out of scope for this page since we already tell the user to do whatever their project prefers. I think the reader typically doesn't have control over which workflow their project uses (they do if it's their own personal project, but then they would typically not use PRs). Still, I agree that it's fair to add a footnote about the disadvantages, just like we have a footnote about the advantages (though I added that footnote only because you pointed out that readers might be confused about our decision).

I agree. Also I think we should leave our footnote, as it's preferred in this project and adds a small explanation.

PhilipMetzger commented on docs/github.md:
```diff
@@ -0,0 +1,90 @@
+# Using Jujutsu with GitHub Projects
```
I rewrote it to: `Using Jujutsu with Github and Gitlab Projects`, since they're similar.

PhilipMetzger commented:
Since the last Revision, I removed both the Git and Mercurial sections and introduced @martinvonz suggestion as basic workflow. PTAL at both the revsets and the basic workflow section. Overall the doc was vastly simplified.

joyously commented on docs/github.md:
````diff
@@ -0,0 +1,90 @@
+is able to create a branch for a revision.
+
+```shell script
+$ ij . 
````
what is `ij`?

PhilipMetzger replied:
IntelliJ. It's a similar feature to vscode's `code .`, which opens the current directory in vscode.

joyously replied:
Please put a comment in the example or use a different command that is well known.

PhilipMetzger replied:
Done, it's now  `code .`

martinvonz replied:
Does `code` start VS Code? I think it's better to not specify a particular code editor here and just have a comment like in the example above.

PhilipMetzger commented on docs/github.md:
```diff
@@ -0,0 +1,90 @@
+$ # Do your work
+$ jj commit
+$ # Jujutsu automatically creates a branch
+$ jj git push --commit $revision
```
Corrected :)

martinvonz reviewed (approved):
I still have a lot of comments, but I think this is a good start. Thanks! Feel free to merge this as is, and then I'll probably go over it and send a follow-up PR later instead.

PhilipMetzger commented on docs/github.md:
```diff
@@ -0,0 +1,130 @@
+
+## Working in a Git co-located repository
+After doing `jj init --git-repo=.`, git will be in a detached HEAD state,
+which is unusual, as git mainly works with branches. 
```
Added

PhilipMetzger commented on docs/github.md:
```diff
@@ -0,0 +1,130 @@
+
+
+Although `jj git import` and `jj git export` exist for co-located repositories,
+they should only be used in case of an unrecoverable error.
```
Removed
